README.md
# Fortune

This is the sources of a [useless website](https://fortune.zedas.fr) that will display a daily result of the `fortune` command display with `cowsay` and `cowsay -f tux`.

[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=Wivik_fortune&metric=alert_status)](https://sonarcloud.io/summary/new_code?id=Wivik_fortune)

## Usage

**Breaking change with the original version, the tool is now fully containerized !**

Pull the container [latest version](https://github.com/Wivik/fortune/pkgs/container/fortune) :

```bash
docker pull ghcr.io/wivik/fortune:latest
```

In a folder of your choice, create a directory name `output` and `chmod` it to 777. That's ugly, but it's just for the container output, you may destroy it after.

Execute the container with a bind mount.

The important part here is the `-w` argument : you need to provide to the tool your website base URL, which is the URL from which the website will be displayed.


```bash
docker run --rm -v ${PWD}/output:/fortune/output ghcr.io/wivik/fortune:latest -w <your website base url>
```

Example in my case, the URL is https://fortune.zedas.fr, I use : `-w https://fortune.zedas.fr`. This URL will be inserted in the template to generate absolute paths.

Then, upload the `output/` content to your website root directory. The content folder will produce the following files :

```bash
./output/index.html
./output/cowsay_feed_atom.xml
./output/tuxsay_feed_atom.xml
./output/cowsay_feed_rss.xml
./output/cowsay.html
./output/tuxsay_feed_rss.xml
./output/tuxsay.html
```

For a daily output, just schedule the script every day at midnight.

## GitHub Actions example

```yaml
name: Publish daily fortune

on:
  workflow_dispatch:
  schedule:
    - cron: '00 00 * * *'

jobs:
  fortune:
    runs-on: ubuntu-latest
    steps:

      - name: Checkout code
        uses: actions/checkout@v3

      - name: Pull image
        run: |
          docker pull ghcr.io/wivik/fortune:latest

      - name: Generate fortune
        run: |
          mkdir output
          chmod 777 output
          docker run --rm -v ${{ github.workspace }}/output:/fortune/output ghcr.io/wivik/fortune:latest -w https://fortune.zedas.fr

      - name: Publish content
        run: |
          cd output
          # use here your publication method, rsync or whatever else, it's up to you.

```

## Under the hood

The first version was a mix of Python and Bash. This one is full Python with a script located in [container/scripts/fortune.py](container/scripts/fortune.py).

This script has only one mandatory param, the website base url.

```bash
usage: fortune.py [-h] -w WEBSITE_URL

options:
  -h, --help            show this help message and exit
  -w WEBSITE_URL, --website-url WEBSITE_URL
                        The website base url : https://fortune.example.com

```




CHANGELOG.md
## [0.2.3](https://github.com/Wivik/fortune/compare/v0.2.2...v0.2.3) (2023-07-31)


### Bug Fixes

* this time the url is good ([c434469](https://github.com/Wivik/fortune/commit/c434469f30ab1c67de6a46eaadc2505aec9f417d))



## [0.2.2](https://github.com/Wivik/fortune/compare/v0.2.1...v0.2.2) (2023-07-31)


### Bug Fixes

* rss files path ([c6bf20f](https://github.com/Wivik/fortune/commit/c6bf20fa64c1b251d2dd302d90a0738c44845171))



## [0.2.1](https://github.com/Wivik/fortune/compare/v0.2.0...v0.2.1) (2023-07-31)


### Bug Fixes

* error in files write ([70428f1](https://github.com/Wivik/fortune/commit/70428f14758cd9f993a71780db067675e98b0339))



# [0.2.0](https://github.com/Wivik/fortune/compare/v0.1.1...v0.2.0) (2023-07-30)


### Features

* add output dir ([8f6cd39](https://github.com/Wivik/fortune/commit/8f6cd39628338d8b2250219036ef0940d3b28d20))



## [0.1.1](https://github.com/Wivik/fortune/compare/v0.1.0...v0.1.1) (2023-07-30)


### Bug Fixes

* security fixes broke the image ([f764c2a](https://github.com/Wivik/fortune/commit/f764c2a6baff3660f9e493db7654126bd073361e))





scripts/fortune.sh
#!/usr/bin/env bash

## run fortune, store the message, and cowsay it in various way

message=$(mktemp)
mypath=$(dirname $0)
day=$(date +"%d")
month=$(date +"%m")
# day="20"
# month="09"
## generate today's fortune

fortune > $message

## we adapt the cowsay file according to a date
case $day$month in
    "0405") cat $message | cowsay -f vader > ${mypath}/cowsay.txt; cat $message | cowsay -f vader > ${mypath}/tuxsay.txt ;;
    "2305") cat $message | cowsay -f turtle > ${mypath}/cowsay.txt; cat $message | cowsay -f turtle > ${mypath}/tuxsay.txt ;;
    "0106") cat $message | cowsay -f supermilker > ${mypath}/cowsay.txt; cat $message | cowsay -f supermilker > ${mypath}/tuxsay.txt ;;
    "0607") cat $message | cowsay -f kiss > ${mypath}/cowsay.txt; cat $message | cowsay -f kiss > ${mypath}/tuxsay.txt ;;
    "0808") cat $message | cowsay -f meow > ${mypath}/cowsay.txt; cat $message | cowsay -f meow > ${mypath}/tuxsay.txt ;;
    "1008") cat $message | cowsay -f moofasa > ${mypath}/cowsay.txt; cat $message | cowsay -f moofasa > ${mypath}/tuxsay.txt ;;
    "2009") cat $message | cowsay -f head-in > ${mypath}/cowsay.txt; cat $message | cowsay -f head-in > ${mypath}/tuxsay.txt ;;
    "3110") cat $message | cowsay -f skeleton > ${mypath}/cowsay.txt; cat $message | cowsay -f skeleton > ${mypath}/tuxsay.txt ;;
    *) cat $message | cowsay > ${mypath}/cowsay.txt; cat $message | cowsay -f tux > ${mypath}/tuxsay.txt ;;
esac

# if [ "$day$month" = "0405" ]; then
#     cat $message | cowsay -f vader > ${mypath}/cowsay.txt
#     cat $message | cowsay -f vader > ${mypath}/tuxsay.txt
# elif [ "$day$month" = "0808" ]; then
#     cat $message | cowsay -f skeleton > ${mypath}/cowsay.txt
#     cat $message | cowsay -f skeleton > ${mypath}/tuxsay.txt
# elif [ "$day$month" = "31" ] && [ "$month" = "10" ]; then
#     cat $message | cowsay -f skeleton > ${mypath}/cowsay.txt
#     cat $message | cowsay -f skeleton > ${mypath}/tuxsay.txt
# else
#     cat $message | cowsay > ${mypath}/cowsay.txt
#     cat $message | cowsay -f tux > ${mypath}/tuxsay.txt
# fi
rm $message

## execute the python script to generate the web pages and the rss feed

/usr/bin/env python3 ${mypath}/fortune.py

## cleanup the mess

rm ${mypath}/cowsay.txt
rm ${mypath}/tuxsay.txt


container/scripts/fortune.py
#!/usr/bin/env python3

from datetime import datetime, time
from feedgen.feed import FeedGenerator
import argparse
import jinja2
import os
import subprocess

## argument
arg_parser = argparse.ArgumentParser()
arg_parser.add_argument('-w', '--website-url', dest='website_url', help='The website base url : https://fortune.example.com', required=True)
args = arg_parser.parse_args()

global website_url
website_url = args.website_url

global script_path
script_path = os.path.dirname(__file__)

global output_dir
output_dir = os.path.join(script_path, 'output')

global fortune_cmd
# fortune_cmd = '/usr/bin/fortune'
fortune_cmd = '/usr/games/fortune'
global cowsay_cmd
# cowsay_cmd = '/usr/bin/cowsay'
cowsay_cmd = '/usr/games/cowsay'


## some vars
# locale_tz = 'Europe/Paris'

special_dates = {
    "0405": "vader",
    "2305": "turtle",
    "0106": "supermilker",
    "0607": "kiss",
    "0808": "meow",
    "1008": "moofasa",
    "2009": "head-in",
    "3110": "skeleton",
}

def fortune():
    """ Run the fortune and cowsay commands according to the day """
    fortune_output = subprocess.check_output(fortune_cmd, shell=True)

    return fortune_output

def is_it_a_special_day():
    """
    function that returns if the day is a special day according to a known list, if not it'll be false
    the event is an option ofr the cowsay command to change the display
    """
    today = datetime.now()
    today_str = today.strftime('%m%d')
    # today_str='1008'
    return special_dates.get(today_str, False)

def cowsay(fortune, special_day=False):
    """
    render the fortune using the cowsay command
    output is a list of two dict because we also have tuxsay
    """
    output = []
    if special_day:
        cowsay_output = subprocess.check_output([cowsay_cmd, '-f', special_day, fortune])
        tuxsay_output = subprocess.check_output([cowsay_cmd, '-f', special_day, fortune])
    else:
        cowsay_output = subprocess.check_output([cowsay_cmd, fortune])
        tuxsay_output = subprocess.check_output([cowsay_cmd, '-f', 'tux', fortune])

    output.append(
        {
            'say': 'cowsay',
            'content': cowsay_output
        }
    )
    output.append(
        {
            'say': 'tuxsay',
            'content': tuxsay_output
        }
    )

    return output


def write_rss(whosays, fortune):
    """ write the rss feed according to the histo folder """
    if whosays == 'cowsay':
        saidby = 'A Cow'
    else:
        saidby = 'Tux'

    feed = FeedGenerator()
    feed.id(website_url)
    feed.title(f'The daily fortune said by {whosays}.')
    feed.description(f'The daily fortune said by {whosays}.')
    feed.author({'name': saidby})
    feed.link(href=website_url, rel='alternate')
    feed.language('en')

    date = datetime.now()
    timestamp = datetime.timestamp(date)
    # print(date)
    # print(int(timestamp))

    fe = feed.add_entry()
    fe.title('Today\'s daily fortune')
    fe.id(f'{website_url}/{whosays}say-{str(int(timestamp))}.html')
    fe.content(f'<pre>{fortune}</pre>')
    fe.link(href=f'{website_url}/{whosays}-say.html')

    feed.atom_file( os.path.join(output_dir, f'{whosays}_feed_atom.xml'))
    feed.rss_file( os.path.join(output_dir, f'{whosays}_feed_rss.xml'))

    # print(feed.atom_str(pretty=True))
    # print(feed.rss_str(pretty=True))


def write_template(fortune, file):
    """ write the fortune page """
    env = jinja2.Environment(loader=jinja2.FileSystemLoader(os.path.join(script_path, 'template')))
    template = env.get_template('index.html.j2')
    render = template.render(fortune=fortune, website_url=website_url, file=file)
    # print(render)
    with open(os.path.join(file), 'w') as f:
        f.write(render)
        f.close()
    print(f'written {os.path.join(file)}')

def main():
    """ main fuction ! """
    ## create output dir if missing
    if not os.path.exists(output_dir):
        print(f'Create output dir {output_dir}')
        os.mkdir(output_dir)
    ## call fortune command
    today_fortune = fortune()
    special_day = is_it_a_special_day()
    ## format cowsay
    says_list = cowsay(today_fortune, special_day)

    ## write html files
    for say in says_list:
        # print(say['say'])
        who_said = say['say']
        say_content = say['content'].decode()
        if say['say'] == 'cowsay':
            # print('cow')
            write_template(say_content, os.path.join(output_dir, 'index.html'))
            write_template(say_content,  os.path.join(output_dir, f'{who_said}.html'))
            write_rss(whosays=who_said, fortune=say_content)
        else:
            # print('tux')
            write_template(say_content,  os.path.join(output_dir, f'{who_said}.html'))
            write_rss(whosays=who_said, fortune=say_content)


if __name__ == '__main__':
    main()

